src/list.c

```diff
@@ -68,3 +107,122 @@ int list_add(struct list_t *list, struct entry_t *entry) {
     previous_node->next = new_node;
     return 0;
 }
+
+/* Função que elimina da lista a entry com a chave key, libertando a
+ * memória ocupada pela entry.
+ * Retorna 0 se encontrou e removeu a entry, 1 se não encontrou a entry,
+ * ou -1 em caso de erro.
+ */
+int list_remove(struct list_t *list, char *key) {
+    // TODO update list size if element was removed
+
+    if (!list || !key) return -1;
+
+    if (!list->head) {
+        return 1;
+    }
+
+    int cmp_ret = strcmp(list->head->entry->key, key);
+    if (!cmp_ret) {
+        list->head = list->head->next;
+        return entry_destroy(list->head->entry);
+    }
+
+    struct node_t* previous_node = list->head;
+    struct node_t* node = previous_node->next;
+
+    while (node != 0) {
+        int cmp_ret = strcmp(node->entry->key, key);
+        if (!cmp_ret) {
+            previous_node->next = node->next;
+            return entry_destroy(list->head->entry);
+        }
+
+        previous_node = node;
+        node = node->next;
+    }
+
+    return 1;
+}
+
+/* Função que obtém da lista a entry com a chave key.
+ * Retorna a referência da entry na lista ou NULL se não encontrar a
+ * entry ou em caso de erro.
+ */
+struct entry_t *list_get(struct list_t *list, char *key) {
+    if (!list || !key) return NULL;
+
+    struct node_t* node = list->head;
+    while (node != 0) {
+        if (strcmp(node->entry->key, key) == 0) {
+            return node->entry;
+        }
+
+        node = node->next;
+    }
+
+    return NULL;
+}
+
+/* Função que conta o número de entries na lista passada como argumento.
+ * Retorna o tamanho da lista ou -1 em caso de erro.
+ */
+int list_size(struct list_t *list) {
+    if (!list) return -1;
+
+    int count = 0;
+    struct node_t* node = list->head;
+    while (node != 0) {
+        count++;
+        node = node->next;
+    }
+
+    return count;
+}
+
+/* Função que constrói um array de char* com a cópia de todas as keys na
+ * lista, colocando o último elemento do array com o valor NULL e
+ * reservando toda a memória necessária.
+ * Retorna o array de strings ou NULL em caso de erro.
+ */
+char **list_get_keys(struct list_t *list) {
```

Comment:
May I ask why is char** (which from my understanding a pointer that points to a char pointer) what we use to represent an array of chars? wouldn't char[]* work to the same intent?

Reply:
> what we use to represent an array of chars?
It's actually an array of an array of chars. From what I understand, I don't think a `C` function can return a `some_type[]`. I've read the other day that static arrays like `int[] number = int[42]`, only live within the parent block where they reside, and get automatically freed when the block ends its execution.

I don't know if that answers your question?

Reply:
![image](https://github.com/padrezulmiro/sd-projeto/assets/1679672/6d85a431-b29d-43ca-97b7-f976a7fdc328)